### PR TITLE
feat: add team distribution registry flow

### PR DESCRIPTION
**Pull request by valerydluski**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Test Case
- [ ] Documentation update
- [ ] Other

#### 🔗 Related issue link

_Describe the source of requirement, like related issue link._

#### 💡 Background and solution

image

#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [ ] Documentation is updated/provided or not needed
- [x] Changes are tested locally

### CODE REVIEW

File: client/src/modules/TeamDistribution/components/TeamDistributionCard/Actions.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import { TeamDistributionDto, TeamDistributionDtoRegistrationStatusEnum } from 'api';
+import { dateWithTimeZoneRenderer } from 'components/Table';
+import moment from 'moment';
+
+const { Text, Link } = Typography;
+
+type Props = {
+  distribution: TeamDistributionDto;
+  onRegister: (distributionId: number) => Promise<void>;
+  onDeleteRegister: (distributionId: number) => Promise<void>;
```
Comment (nataliagulko) [inline client/src/modules/TeamDistribution/components/TeamDistributionCard/Actions.tsx:13]: I would prefer to name methods without "on"
```
  register: (distributionId: number) => Promise<void>;
  deleteRegistration: (distributionId: number) => Promise<void>;
```

File: client/src/modules/TeamDistribution/components/TeamDistributionCard/Actions.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+export function Actions({ distribution, onRegister, onDeleteRegister }: Props) {
+  const timezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
+
+  const handleCancel = async () => {
+    Modal.confirm({
+      title: 'Cancel registration',
+      content: (
+        <>
+          Are you sure you want to cancel your group task registration? You will be able to register again until{' '}
+          {dateWithTimeZoneRenderer(timezone, 'YYYY-MM-DD HH:mm')(distribution.endDate)}.
```
Comment (nataliagulko) [inline client/src/modules/TeamDistribution/components/TeamDistributionCard/Actions.tsx:34]: Seems like you can extract it to separate function or variable and reuse it below

File: client/src/modules/TeamDistribution/components/TeamDistributionCard/Actions.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+          <Space size={24}>
+            <Button disabled>Register</Button>
+            <Text type="secondary">Registration is closed</Text>
+          </Space>
+        </Row>
+      );
+
+    default:
+      return null;
+  }
```
Comment (nataliagulko) [inline client/src/modules/TeamDistribution/components/TeamDistributionCard/Actions.tsx:103]: Seems like you can extract Row + Space into single return statement and in switch change only Button + Text properties/texts

File: client/src/modules/TeamDistribution/components/TeamDistributionCard/TeamDistributionCard.test.tsx [hunk 4]
```diff
@@ -45,7 +51,14 @@ describe('TeamDistributionCard', () => {
...
     render(
-      <TeamDistributionCard distribution={distribution} isManager={true} onDelete={mockOnDelete} onEdit={mockOnEdit} />,
+      <TeamDistributionCard
+        distribution={distribution}
+        isManager={true}
+        onDelete={mockOnDelete}
+        onEdit={mockOnEdit}
+        onDeleteRegister={onDeleteRegister}
+        onRegister={onRegister}
+      />,
```
Comment (nataliagulko) [inline client/src/modules/TeamDistribution/components/TeamDistributionCard/TeamDistributionCard.test.tsx:61]: You can extract the entire render(<TeamDistrubutionCard ... />) into separate function to change props in one place if needed. Check example client\src\modules\AutoTest\components\Coding\Coding.test.tsx

File: nestjs/src/courses/students/students.service.ts [hunk 5]
```diff
@@ -15,6 +16,37 @@ export class StudentsService {
...
+      where: { id: studentId },
+      relations: ['teamDistribution'],
+    });
+    return student;
+  }
+
+  public async addStudentToTeamDistribution(studentId: number, teamDistribution: TeamDistribution) {
+    const student = await this.getStudentWithTeamDistributions(studentId);
+    const currTimestampUTC = Date.now();
+    const distributionStartDate = teamDistribution.startDate.getTime();
```
Comment (nataliagulko) [inline nestjs/src/courses/students/students.service.ts:30]: Please consider to use dayjs utilities

**Review by nataliagulko — APPROVED** [review]

**Comment by valerydluski** [comment]
@nataliagulko thank you for the review, I have corrected all the comments.

**Pull request merged**